File: go/kernel/authority.go
package kernel

import (
	"slices"

	"google.golang.org/protobuf/proto"

	pb "platformkernel/gen/platform/kernel/v1alpha1"
)

// Authorities implements K5 (Contract/spec/K5-authority.md): declarations, the
// receiver's check, and the outbox of one edge.
type Authorities struct {
	edge    string
	current map[[2]string]*pb.AuthorityDeclaration // (tenant, data class)
	outbox  map[[2]string]*outboxEntry             // (tenant, idempotency key)
}

type outboxEntry struct {
	submission *pb.Submission
	state      pb.SubmissionState
}

// transitions lists, per event, the allowed from → to states (A5).
var transitions = map[string]map[pb.SubmissionState]pb.SubmissionState{
	"send":     {pb.SubmissionState_SUBMISSION_STATE_PENDING: pb.SubmissionState_SUBMISSION_STATE_SENDING},
	"confirm":  {pb.SubmissionState_SUBMISSION_STATE_SENDING: pb.SubmissionState_SUBMISSION_STATE_CONFIRMED},
	"conflict": {pb.SubmissionState_SUBMISSION_STATE_SENDING: pb.SubmissionState_SUBMISSION_STATE_CONFLICT},
	"reject":   {pb.SubmissionState_SUBMISSION_STATE_SENDING: pb.SubmissionState_SUBMISSION_STATE_REJECTED},
	"timeout":  {pb.SubmissionState_SUBMISSION_STATE_SENDING: pb.SubmissionState_SUBMISSION_STATE_UNKNOWN},
	"retry":    {pb.SubmissionState_SUBMISSION_STATE_UNKNOWN: pb.SubmissionState_SUBMISSION_STATE_SENDING},
}

func NewAuthorities(edge string) *Authorities {
	return &Authorities{edge: edge, current: map[[2]string]*pb.AuthorityDeclaration{}, outbox: map[[2]string]*outboxEntry{}}
}

func (a *Authorities) Declare(d *pb.AuthorityDeclaration) *Error {
	if slices.Contains([]string{d.GetTenantId(), d.GetDataClass(), d.GetAuthorityId()}, "") ||
		d.GetKind() == pb.AuthorityKind_AUTHORITY_KIND_UNSPECIFIED {
		return errorf(pb.ErrorCode_ERROR_CODE_INVALID_ARGUMENT) // A1
	}
	key := [2]string{d.GetTenantId(), d.GetDataClass()}
	if d.GetEpoch() != a.current[key].GetEpoch()+1 {
		return errorf(pb.ErrorCode_ERROR_CODE_CONFLICT) // A2
	}
	a.current[key] = d
	return nil
}

func (a *Authorities) declaration(s *pb.Submission) (*pb.AuthorityDeclaration, *Error) {
	d := a.current[[2]string{s.GetTenantId(), s.GetTarget().GetType()}]
	if d == nil {
		return nil, errorf(pb.ErrorCode_ERROR_CODE_NOT_FOUND)
	}
	return d, nil
}

// Authorize is the receiver's check before K4 accepts a submission (A3).
func (a *Authorities) Authorize(s *pb.Submission) *Error {
	d, err := a.declaration(s)
	if err == nil && s.GetAuthority() != d.GetAuthorityId() {
		err = errorf(pb.ErrorCode_ERROR_CODE_NOT_AUTHORITY)
	}
	return err
}

// Enqueue puts a submission in this edge's outbox (A4, A6).
func (a *Authorities) Enqueue(s *pb.Submission) (pb.SubmissionState, *Error) {
	d, err := a.declaration(s)
	if err != nil {
		return 0, err
	}
	key := [2]string{s.GetTenantId(), s.GetIdempotencyKey()}
	if existing := a.outbox[key]; existing != nil {
		if !proto.Equal(existing.submission, s) {
			return 0, errorf(pb.ErrorCode_ERROR_CODE_IDEMPOTENCY_CONFLICT)
		}
		return existing.state, nil
	}
	state := pb.SubmissionState_SUBMISSION_STATE_PENDING
	if d.GetAuthorityId() == a.edge {
		state = pb.SubmissionState_SUBMISSION_STATE_CONFIRMED
	}
	a.outbox[key] = &outboxEntry{submission: s, state: state}
	return state, nil
}

func (a *Authorities) Transition(tenant, idempotencyKey, event string) (pb.SubmissionState, *Error) {
	entry := a.outbox[[2]string{tenant, idempotencyKey}]
	if entry == nil {
		return 0, errorf(pb.ErrorCode_ERROR_CODE_NOT_FOUND)
	}
	next, ok := transitions[event][entry.state]
	if !ok {
		return 0, errorf(pb.ErrorCode_ERROR_CODE_INVALID_ARGUMENT) // A5
	}
	entry.state = next
	return next, nil
}

File: swift/Sources/KernelModel/Authority.swift
/// K5 Authority and sync (Contract/spec/K5-authority.md).
public enum AuthorityKind: String, Codable, Sendable {
    case unspecified = "AUTHORITY_KIND_UNSPECIFIED"
    case device = "AUTHORITY_KIND_DEVICE"
    case tenantServer = "AUTHORITY_KIND_TENANT_SERVER"
    case external = "AUTHORITY_KIND_EXTERNAL"
}

public struct AuthorityDeclaration: Hashable, Codable, Sendable {
    public var tenantId = ""
    public var dataClass = ""
    public var kind = AuthorityKind.unspecified
    public var authorityId = ""
    public var epoch: UInt32 = 0

    public init() {}

    // Proto3 JSON omits default values.
    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        tenantId = try c.decodeIfPresent(String.self, forKey: .tenantId) ?? ""
        dataClass = try c.decodeIfPresent(String.self, forKey: .dataClass) ?? ""
        kind = try c.decodeIfPresent(AuthorityKind.self, forKey: .kind) ?? .unspecified
        authorityId = try c.decodeIfPresent(String.self, forKey: .authorityId) ?? ""
        epoch = try c.decodeIfPresent(UInt32.self, forKey: .epoch) ?? 0
    }
}

public enum SubmissionState: String, Codable, Sendable {
    case unspecified = "SUBMISSION_STATE_UNSPECIFIED"
    case pending = "SUBMISSION_STATE_PENDING"
    case sending = "SUBMISSION_STATE_SENDING"
    case confirmed = "SUBMISSION_STATE_CONFIRMED"
    case conflict = "SUBMISSION_STATE_CONFLICT"
    case rejected = "SUBMISSION_STATE_REJECTED"
    case unknown = "SUBMISSION_STATE_UNKNOWN"
}

public enum OutboxEvent: String, Sendable {
    case send, confirm, conflict, reject, timeout, retry

    /// The allowed from → to states (A5).
    var transitions: [SubmissionState: SubmissionState] {
        switch self {
        case .send: [.pending: .sending]
        case .confirm: [.sending: .confirmed]
        case .conflict: [.sending: .conflict]
        case .reject: [.sending: .rejected]
        case .timeout: [.sending: .unknown]
        case .retry: [.unknown: .sending]
        }
    }
}

/// Declarations, the receiver's check, and the outbox of one edge.
public struct Authorities: Sendable {
    private struct Key: Hashable { let tenant, name: String }
    private let edge: String
    private var current: [Key: AuthorityDeclaration] = [:]                         // (tenant, data class)
    private var outbox: [Key: (submission: Submission, state: SubmissionState)] = [:] // (tenant, idempotency key)

    public init(edge: String) {
        self.edge = edge
    }

    public mutating func declare(_ d: AuthorityDeclaration) throws(KernelError) {
        guard ![d.tenantId, d.dataClass, d.authorityId].contains(where: \.isEmpty), d.kind != .unspecified else {
            throw .invalidArgument                                                                 // A1
        }
        let key = Key(tenant: d.tenantId, name: d.dataClass)
        guard d.epoch == (current[key]?.epoch ?? 0) + 1 else { throw .conflict }                   // A2
        current[key] = d
    }

    private func declaration(for s: Submission) throws(KernelError) -> AuthorityDeclaration {
        guard let d = current[Key(tenant: s.tenantId, name: s.target.type)] else { throw .notFound }
        return d
    }

    /// The receiver's check before K4 accepts a submission (A3).
    public func authorize(_ s: Submission) throws(KernelError) {
        guard try declaration(for: s).authorityId == s.authority else { throw .notAuthority }
    }

    /// Puts a submission in this edge's outbox (A4, A6).
    public mutating func enqueue(_ s: Submission) throws(KernelError) -> SubmissionState {
        let d = try declaration(for: s)
        let key = Key(tenant: s.tenantId, name: s.idempotencyKey)
        if let existing = outbox[key] {
            guard existing.submission == s else { throw .idempotencyConflict }
            return existing.state
        }
        let state: SubmissionState = d.authorityId == edge ? .confirmed : .pending
        outbox[key] = (s, state)
        return state
    }

    public mutating func transition(tenant: String, idempotencyKey: String, event: String) throws(KernelError) -> SubmissionState {
        let key = Key(tenant: tenant, name: idempotencyKey)
        guard let entry = outbox[key] else { throw .notFound }
        guard let next = OutboxEvent(rawValue: event)?.transitions[entry.state] else { throw .invalidArgument } // A5
        outbox[key]?.state = next
        return next
    }
}
